test/e2e/service.go

```diff
@@ -226,6 +226,74 @@ var _ = framework.KubeDescribe("Services", func() {
 		validateEndpointsOrFail(c, ns, serviceName, PortsByPodName{})
 	})
 
+	// TODO: verify source IP preservation for LoadBalancer type services when applicable
+	It("should preserve source pod IP for traffic thru service cluster IP", func() {
+
+		serviceName := "sourceip-test"
+		ns := f.Namespace.Name
+
+		By("creating a TCP service " + serviceName + " with type=ClusterIP in namespace " + ns)
+		jig := NewServiceTestJig(c, serviceName)
+		servicePort := 8080
+		tcpService := jig.CreateTCPServiceWithPort(ns, nil, int32(servicePort))
+		jig.SanityCheckService(tcpService, api.ServiceTypeClusterIP)
+		defer func() {
+			By("Cleaning up the sourceip test service")
+			err := c.Services(ns).Delete(serviceName)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+		serviceIp := tcpService.Spec.ClusterIP
+		framework.Logf("sourceip-test cluster ip: %s", serviceIp)
+
+		By("Picking multiple nodes")
+		nodes := framework.GetReadySchedulableNodesOrDie(f.Client)
+
+		if len(nodes.Items) == 1 {
+			framework.Skipf("The test requires two Ready nodes on %s, but found just one.", framework.TestContext.Provider)
+		}
+
+		node1 := nodes.Items[0]
+		node2 := nodes.Items[1]
+
+		By("Creating a webserver pod be part of the TCP service which echoes back source ip")
+		serverPodName := "echoserver-sourceip"
+		jig.launchEchoserverPodOnNode(f, node1.Name, serverPodName)
+		defer func() {
+			By("Cleaning up the echo server pod")
+			err := c.Pods(ns).Delete(serverPodName, nil)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+
+		// Waiting for service to expose endpoint
+		validateEndpointsOrFail(c, ns, serviceName, PortsByPodName{serverPodName: {servicePort}})
+
+		By("Creating an exec pod on another node")
+		execPodName := createExecPodOnNode(f.Client, ns, node2.Name, "execpod-sourceip")
+		defer func() {
+			By("Cleaning up the exec pod")
+			err := c.Pods(ns).Delete(execPodName, nil)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+		podClient := f.Client.Pods(ns)
+		execPod, err := podClient.Get(execPodName)
+		ExpectNoError(err)
+		execPodIp := execPod.Status.PodIP
+		framework.Logf("exec pod ip: %s", execPodIp)
+
+		By("Getting echo response from service")
+		cmd := fmt.Sprintf(`wget -T 30 -qO- %s:%d | grep client_address | cut -f 2 -d "="`, serviceIp, servicePort)
+		sourceIp, err := framework.RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		if err != nil {
+			framework.Failf("error while kubectl execing %q in pod %v/%v: %v\nOutput: %v", cmd, execPod.Namespace, execPod.Name, err, sourceIp)
+		}
+		// the stdout return from RunHostCmd seems to come with "\n", so TrimSpace is needed
+		sourceIp = strings.TrimSpace(sourceIp)
+		framework.Logf("Echoed source ip: %s", sourceIp)
```

Comment:
nit: I think you can skip this. Because the Expect statement under this will shout if things went wrong.